### PR TITLE
Add required_mod_options

### LINKED ISSUE #114: Add `required_mod_options` and raise when `mod_options` are missing them
`model_contract_method_name`
`pipeline_idx`

### PR DESCRIPTION
**Pull request by gabegma**
Resolve #114 

## Description:
- When calling a module without mandatory mod_options, it would sometimes fail silently, or the error message would be hard to understand. This proposal makes it clear when specific mod_options are required.

## Checklist:

You should check all boxes before the PR is ready. If a box does not apply, check it to acknowledge it.

* [x] **ISSUE NUMBER.** You linked the issue number (Ex: Resolve #XXX).
* [x] **PRE-COMMIT.** You ran pre-commit on all commits, or else, you
  ran `pre-commit run --all-files` at the end.
* [x] **USER CHANGES.** The changes are added to CHANGELOG.md and the documentation, if they impact
  our users.
* [x] **DEV CHANGES.**
    * Update the documentation if this PR changes how to develop/launch on the app.
    * Update the `README` files and our wiki for any big design decisions, if relevant.
    * Add unit tests, docstrings, typing and comments for complex sections.

### CODE REVIEW

File: azimuth/modules/base_classes/module.py [hunk 1]
```diff
@@ -31,11 +32,11 @@ def __init__(
...
         mod_options = mod_options or ModuleOptions()
         self.mod_options = mod_options
-        if diff := (
-            set(self.mod_options.no_alias_dict(exclude_defaults=True).keys())
-            - self.allowed_mod_options
-        ):
+        defined_mod_options = set(self.mod_options.no_alias_dict(exclude_defaults=True).keys())
+        if diff := (defined_mod_options - self.allowed_mod_options - self.required_mod_options):
             raise ValueError(f"Unexpected mod_options {diff} for {self.__class__.__name__}.")
+        if self.required_mod_options and self.required_mod_options.isdisjoint(defined_mod_options):
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/module.py:38]: I think we should check if ALL options in `required_mod_options` are in `required_mod_options`. We can use `issubset` or `<=`.
```suggestion
        if not self.required_mod_options.issubset(defined_mod_options):
```
The `isdisjoint` will pass as soon as ONE option in `required_mod_options` is in `required_mod_options`. For example, here, `2` is missing.
```python
>>> {1, 2}.isdisjoint({1})
False
```

Comment (gabegma) [reply azimuth/modules/base_classes/module.py:38]: Oh, right! Thank you!

File: azimuth/modules/base_classes/module.py [hunk 2]
```diff
@@ -52,7 +53,8 @@ def get_effective_arguments(self) -> ModuleEffectiveArguments:
         # model_contract_method_name are excluded too because it's already in the task_name.
         return ModuleEffectiveArguments(
             mod_options=self.mod_options.dict(
-                exclude={"indices", "model_contract_method_name"}, include=self.allowed_mod_options
+                exclude={"indices", "model_contract_method_name"},
+                include=self.allowed_mod_options.union(self.required_mod_options),
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/module.py:57]: I see in most places you used the operator `|` for the union. I don't have a preference for one or for the other, but I guess we should keep it consistent.
```suggestion
                include=self.allowed_mod_options | self.required_mod_options,
```

**Review by JosephMarinier — APPROVED** [review]
Good catch! 👍 Thank you!

File: azimuth/modules/base_classes/module.py (hunk 2 above)
Comment (gabegma) [reply azimuth/modules/base_classes/module.py:57]: Right!! I forgot about it. Thanks!

File: azimuth/modules/README.md [hunk 3]
```diff
@@ -76,9 +77,8 @@ inherit from. Different options are available, with different particularities.
 - It inherits from Dataset Result Module, and is a module that wraps models.
 - These modules handle all the tasks which are in need of direct access to the model in order to
   compute their results (e.g. prediction, saliency).
-- `allowed_mod_options` include some default options that can affect models, such as the `threshold`
-  and the `pipeline_index`. The mod_options `model_contract_method_name` is mandatory as it
-  specifies which method should be called in the module.
+- `allowed_mod_options` include default options that can affect models, such as the `threshold`.
```
Comment (JosephMarinier) [inline azimuth/modules/README.md:80]: I'm not sure I get what we mean by "default" options.

File: azimuth/modules/base_classes/module.py [hunk 4]
```diff
@@ -22,6 +22,7 @@ class Module(DaskModule[ConfigScope]):
     with the dataset and the pipelines."""
 
     allowed_mod_options: Set[str] = set()
```
Comment (JosephMarinier) [inline azimuth/modules/base_classes/module.py:24]: To make it clear that those don't also include the required options (in other words, they are disjoint), do you think `optional_mod_options` would be better? Or would it be a pain to rename?

Comment (gabegma) [reply azimuth/modules/base_classes/module.py:24]: Yeah good idea, I can rename.

File: azimuth/modules/README.md (hunk 3 above)
Comment (gabegma) [reply azimuth/modules/README.md:80]: I just removed it in my new PR.

File: azimuth/modules/base_classes/module.py (hunk 4 above)
Comment (gabegma) [reply azimuth/modules/base_classes/module.py:24]: See PR #416

**Pull request merged**